### PR TITLE
containerd/cri/integration-tests.sh: Add TestContainerMemoryUpdate

### PR DESCRIPTION
**Pull request by teawater**
Add TestContainerMemoryUpdate to test the memory update function with
or without virtio-mem.

Fixes: #3508

Signed-off-by: Hui Zhu <teawater@antfin.com>

### CODE REVIEW

File: integration/containerd/cri/integration-tests.sh [hunk 1]
```diff
@@ -213,13 +225,58 @@ EOF
...
 }
 
+TestContainerMemoryUpdate() {
+	if [[ "${KATA_HYPERVISOR}" != "qemu" ]]; then
+		return
+	fi
+
+	test_virtio_mem=$1
+
+	if [ $test_virtio_mem ]; then
```
Comment (liubin) [inline integration/containerd/cri/integration-tests.sh:241]: normally `if [ -n $str_exceptec_not_null ]; ` ?

File: integration/containerd/cri/integration-tests.sh [hunk 2]
```diff
@@ -213,13 +225,58 @@ EOF
...
+	testContainerStart
+
+	vmSize=$(($(crictl exec -t -i $c cat /proc/meminfo | grep "MemTotal:" | awk '{print $2}')*1024))
+	if [ $vmSize -gt $((2*1024*1024*1024)) ] || [ $vmSize -lt $((2*1024*1024*1024-128*1024*1024)) ]; then
+		die "The VM memory size $vmSize before update is not right"
+	fi
+
+	sudo crictl update --memory $((2*1024*1024*1024)) $cid
+
+	vmSize=$(($(crictl exec -t -i $c cat /proc/meminfo | grep "MemTotal:" | awk '{print $2}')*1024))
```
Comment (liubin) [inline integration/containerd/cri/integration-tests.sh:263]: `vmSize` -> `vm_size`

File: integration/containerd/cri/integration-tests.sh [hunk 3]
```diff
@@ -213,13 +225,58 @@ EOF
...
+	if [ $vmSize -gt $((4*1024*1024*1024)) ] || [ $vmSize -lt $((4*1024*1024*1024-128*1024*1024)) ]; then
+		die "The VM memory size $vmSize after increase is not right"
+	fi
+
+	if [ $test_virtio_mem ]; then
+		sudo crictl update --memory $((1*1024*1024*1024)) $cid
+
+		vmSize=$(($(crictl exec -t -i $c cat /proc/meminfo | grep "MemTotal:" | awk '{print $2}')*1024))
+		if [ $vmSize -gt $((3*1024*1024*1024)) ] || [ $vmSize -lt $((3*1024*1024*1024-128*1024*1024)) ]; then
+			die "The VM memory size $vmSize after decrease is not right"
```
Comment (liubin) [inline integration/containerd/cri/integration-tests.sh:273]: Should ensure the pod to be deleted(by calling `testContainerStop`) even if the assertion is failed.

File: integration/containerd/cri/integration-tests.sh (hunk 1 above)
Comment (teawater) [reply integration/containerd/cri/integration-tests.sh]: Updated.

Comment (teawater) [reply integration/containerd/cri/integration-tests.sh]: Updated.

File: integration/containerd/cri/integration-tests.sh (hunk 2 above)
Comment (teawater) [reply integration/containerd/cri/integration-tests.sh]: Updated.

File: integration/containerd/cri/integration-tests.sh (hunk 3 above)
Comment (teawater) [reply integration/containerd/cri/integration-tests.sh]: Fixed.

File: integration/containerd/cri/integration-tests.sh [hunk 4]
```diff
@@ -287,6 +347,9 @@ main() {
 			make -e test-integration
 	done
 
+	TestContainerMemoryUpdate 1
+	TestContainerMemoryUpdate
```
Comment (c3d) [inline integration/containerd/cri/integration-tests.sh:351]: Nit: for consistency, I would have used 1 and 0.

**Review by c3d — APPROVED** [review]
Looks OK, one minor comment.

File: integration/containerd/cri/integration-tests.sh (hunk 4 above)
Comment (teawater) [reply integration/containerd/cri/integration-tests.sh]: Updated.

**Comment by teawater** [comment]
> lgtm, thanks @teawater!
> 
> I also would like to see this one backported to the `stable-2.1` branch.

Done.

**Comment by liubin** [comment]
I added `do-not-merge` to prevent this pr be merged, http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu-initrd-PR/602/console failed that run tests in file `integration/containerd/cri/integration-tests.sh`.
```

**Comment by fidencio** [comment]
@teawater, the error on the ARM CI is real and is related to your test.  Please, take a look whenever you find some time.
```
16:40:39 ./hack/test-utils.sh: line 82: 74058 Terminated              keepalive "sudo PATH=${PATH} ${ROOT}/_output/${CONTAINERD_BIN} ${CONTAINERD_FLAGS}" ${RESTART_WAIT_PERIOD} &> ${report_dir}/containerd.log
16:40:39 make[1]: Leaving directory '/home/jenkins/workspace/kata-containers-2.0-tests-ubuntu-ARM-PR/go/src/github.com/containerd/cri'
16:40:39 INFO: Test container memory update without virtio-mem
16:40:39 Image is up to date for sha256:f6467c4e9e1526a6e856444fde786794014c628ab8d64a2eaca5ca1a95ff13de
16:40:46 cd51c67105ec4ab8ad2a8e8d6c52439cc72c97a072c24dd2bb01031aab54c486
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Create stream
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream added, broadcasting: 1
16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 1
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Create stream
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Stream added, broadcasting: 3
16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 3
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Create stream
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Stream added, broadcasting: 5
16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 5
16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 3
16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame handling
16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame sent
16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 3
16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame handling
16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 5
16:40:53 2021/05/12 14:40:53 (0x4000256320) (5) Data frame handling
16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 1
16:40:53 2021/05/12 14:40:53 (0x400033a140) (1) Data frame handling
16:40:53 2021/05/12 14:40:53 (0x400033a140) (1) Data frame sent
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream removed, broadcasting: 1
16:40:53 2021/05/12 14:40:53 (0x400035c000) Go away received
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream removed, broadcasting: 1
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Stream removed, broadcasting: 3
16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Stream removed, broadcasting: 5
16:40:53 cd51c67105ec4ab8ad2a8e8d6c52439cc72c97a072c24dd2bb01031aab54c486
16:40:54 2021/05/12 14:40:54 (0x40002bf8c0) (0x400038c460) Create stream
16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400038c460) Stream added, broadcasting: 1
16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) Reply frame received for 1
16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400024e0a0) Create stream
16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400024e0a0) Stream added, broadcasting: 3
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Reply frame received for 3
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Create stream
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Stream added, broadcasting: 5
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Reply frame received for 5
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 5
16:40:55 2021/05/12 14:40:55 (0x400024e140) (5) Data frame handling
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 1
16:40:55 2021/05/12 14:40:55 (0x400038c460) (1) Data frame handling
16:40:55 2021/05/12 14:40:55 (0x400038c460) (1) Data frame sent
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400038c460) Stream removed, broadcasting: 1
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 3
16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame handling
16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame sent
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 3
16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame handling
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Go away received
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400038c460) Stream removed, broadcasting: 1
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e0a0) Stream removed, broadcasting: 3
16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Stream removed, broadcasting: 5
16:40:55 INFO: stop pod d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
16:40:55 Stopped sandbox d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
16:40:55 INFO: remove pod d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
16:40:55 Removed sandbox d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
16:40:55 ERROR: The VM memory size 2094370816 after increase is not right
16:40:55 Makefile:144: recipe for target 'cri-containerd' failed
16:40:55 make: *** [cri-containerd] Error 1
```

**Comment by teawater** [comment]
> @teawater, the error on the ARM CI is real and is related to your test. Please, take a look whenever you find some time.
> 
> ```
> 16:40:39 ./hack/test-utils.sh: line 82: 74058 Terminated              keepalive "sudo PATH=${PATH} ${ROOT}/_output/${CONTAINERD_BIN} ${CONTAINERD_FLAGS}" ${RESTART_WAIT_PERIOD} &> ${report_dir}/containerd.log
> 16:40:39 make[1]: Leaving directory '/home/jenkins/workspace/kata-containers-2.0-tests-ubuntu-ARM-PR/go/src/github.com/containerd/cri'
> 16:40:39 INFO: Test container memory update without virtio-mem
> 16:40:39 Image is up to date for sha256:f6467c4e9e1526a6e856444fde786794014c628ab8d64a2eaca5ca1a95ff13de
> 16:40:46 cd51c67105ec4ab8ad2a8e8d6c52439cc72c97a072c24dd2bb01031aab54c486
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Create stream
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream added, broadcasting: 1
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 1
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Create stream
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Stream added, broadcasting: 3
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 3
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Create stream
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Stream added, broadcasting: 5
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Reply frame received for 5
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 3
> 16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame handling
> 16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame sent
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 3
> 16:40:53 2021/05/12 14:40:53 (0x40005ae000) (3) Data frame handling
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 5
> 16:40:53 2021/05/12 14:40:53 (0x4000256320) (5) Data frame handling
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Data frame received for 1
> 16:40:53 2021/05/12 14:40:53 (0x400033a140) (1) Data frame handling
> 16:40:53 2021/05/12 14:40:53 (0x400033a140) (1) Data frame sent
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream removed, broadcasting: 1
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) Go away received
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x400033a140) Stream removed, broadcasting: 1
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x40005ae000) Stream removed, broadcasting: 3
> 16:40:53 2021/05/12 14:40:53 (0x400035c000) (0x4000256320) Stream removed, broadcasting: 5
> 16:40:53 cd51c67105ec4ab8ad2a8e8d6c52439cc72c97a072c24dd2bb01031aab54c486
> 16:40:54 2021/05/12 14:40:54 (0x40002bf8c0) (0x400038c460) Create stream
> 16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400038c460) Stream added, broadcasting: 1
> 16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) Reply frame received for 1
> 16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400024e0a0) Create stream
> 16:40:55 2021/05/12 14:40:54 (0x40002bf8c0) (0x400024e0a0) Stream added, broadcasting: 3
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Reply frame received for 3
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Create stream
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Stream added, broadcasting: 5
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Reply frame received for 5
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 5
> 16:40:55 2021/05/12 14:40:55 (0x400024e140) (5) Data frame handling
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 1
> 16:40:55 2021/05/12 14:40:55 (0x400038c460) (1) Data frame handling
> 16:40:55 2021/05/12 14:40:55 (0x400038c460) (1) Data frame sent
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400038c460) Stream removed, broadcasting: 1
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 3
> 16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame handling
> 16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame sent
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Data frame received for 3
> 16:40:55 2021/05/12 14:40:55 (0x400024e0a0) (3) Data frame handling
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) Go away received
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400038c460) Stream removed, broadcasting: 1
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e0a0) Stream removed, broadcasting: 3
> 16:40:55 2021/05/12 14:40:55 (0x40002bf8c0) (0x400024e140) Stream removed, broadcasting: 5
> 16:40:55 INFO: stop pod d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
> 16:40:55 Stopped sandbox d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
> 16:40:55 INFO: remove pod d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
> 16:40:55 Removed sandbox d62f23a8e051f3a37cb04e85b1a77d3986021f01413b856d132230b456b0222a
> 16:40:55 ERROR: The VM memory size 2094370816 after increase is not right
> 16:40:55 Makefile:144: recipe for target 'cri-containerd' failed
> 16:40:55 make: *** [cri-containerd] Error 1
> ```

Looks hotplug memory is not work OK in aarch64.  I posted a issue https://github.com/kata-containers/kata-containers/issues/1841 and I will add code to let the memory hotplug test just run with x86_64.

**Comment by teawater** [comment]
> I added `do-not-merge` to prevent this pr be merged, http://jenkins.katacontainers.io/job/kata-containers-2.0-tests-ubuntu-initrd-PR/602/console failed that run tests in file `integration/containerd/cri/integration-tests.sh`.

@liubin This issue is fixed.  Could you remove this `do-not-merge`?

**Pull request merged**